packages/gatsby/src/db/common/query.ts

```diff
@@ -0,0 +1,211 @@
+        })
+      } else if (isDbComparator(key)) {
+        return [
+          {
+            type: `query`,
+            path,
+            query: {
+              comparator: key,
+              value: filter[key],
+            },
+          },
+        ]
+      } else {
+        return createDbQueriesFromObjectNested(filter[key], path.concat([key]))
+      }
+    }
+  )
+}
+
+export function prefixResolvedFields(
+  queries: Array<DbQuery>,
+  resolvedFields: object
+): Array<DbQuery> {
+  const dottedFields = objectToDottedField(resolvedFields)
+  const dottedFieldKeys = Object.getOwnPropertyNames(dottedFields)
+  queries.forEach(query => {
+    const prefixPath = query.path.join(`.`)
+    if (
+      dottedFields[prefixPath] ||
+      (dottedFieldKeys.some(dottedKey => dottedKey.startsWith(prefixPath)) &&
+        query.type === `elemMatch`) ||
+      dottedFieldKeys.some(dottedKey => prefixPath.startsWith(dottedKey))
+    ) {
+      query.path.unshift(`__gatsby_resolved`)
+    }
+  })
+  return queries
+}
+
+export function dbQueryToSiftQuery(query: DbQuery): object {
+  const result = {}
+  if (query.type === `elemMatch`) {
+    result[query.path.join(`.`)] = {
+      $elemMatch: dbQueryToSiftQuery(query.nestedQuery),
+    }
+  } else if (query.path.length) {
+    result[query.path.join(`.`)] = {
+      [query.query.comparator]: query.query.value,
+    }
+  } else {
+    return {
+      [query.query.comparator]: query.query.value,
+    }
+  }
+  return result
+}
+
+// Most of the below can be gone after we decide to remove loki
+
+// Converts a nested mongo args object into a dotted notation. acc
+// (accumulator) must be a reference to an empty object. The converted
+// fields will be added to it. E.g
+//
+// {
+//   internal: {
+//     type: {
+//       $eq: "TestNode"
+//     },
+//     content: {
+//       $regex: new MiniMatch(v)
+//     }
+//   },
+//   id: {
+//     $regex: newMiniMatch(v)
+//   }
+// }
+//
+// After execution, acc would be:
+//
+// {
+//   "internal.type": {
+//     $eq: "TestNode"
+//   },
+//   "internal.content": {
+//     $regex: new MiniMatch(v)
+//   },
+//   "id": {
+//     $regex: // as above
+//   }
+// }
+export function toDottedFields(filter, acc = {}, path = []): object {
+  Object.keys(filter).forEach(key => {
+    const value = filter[key]
+    const nextValue = _.isPlainObject(value) && value[Object.keys(value)[0]]
+    if (key === `$elemMatch`) {
+      acc[path.join(`.`)] = { [`$elemMatch`]: toDottedFields(value) }
+    } else if (_.isPlainObject(nextValue)) {
+      toDottedFields(value, acc, path.concat(key))
+    } else {
+      acc[path.concat(key).join(`.`)] = value
+    }
+  })
+  return acc
+}
+
+// Like above, but doesn't handle $elemMatch
+export function objectToDottedField(obj, path = []): object {
+  let result = {}
+  Object.keys(obj).forEach(key => {
+    const value = obj[key]
+    if (_.isPlainObject(value)) {
+      const pathResult = objectToDottedField(value, path.concat(key))
```

Comment:
`path.concat(key)`? I guess either `path.push(key)` or `path.join(key)` or `path.concat(value)` but I'm not even sure what's going on here. Same for the other one. (TS is complaining about this).
tip: `yarn typecheck`

Reply:
I think it expects a typing for `path: Array<string>` in function signature 🤷‍♂